### ISSUE TITLE
Unexpected nullable warning due to null-conditional access

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#34942]
**Version Used**:
3.1.0-beta1-19172-05+edd2de88fb3e84a097fb30b4070e0f219f624e40

**Steps to Reproduce**:
```C#
#nullable enable
using System;

public class Program
{
    static void Main()
    {
        string? value = "42";
        int count = 84;
        if (value?.Length == count)
        {
            Console.WriteLine(value.Length); // warning
        }
    }
}
```

Even though value was explicitly set to a non-null value, and even though the `if (value?.Length == count)` effectively expands to `if (value != null && value.Length == count)` and we could only end up inside the `if` block if `value` is non-null, this still warns with "Possible dereference of a null reference." on `value.Length`.

Interestingly, changing the `count` to be a const makes it no longer warn:
```C#
#nullable enable
using System;

public class Program
{
    static void Main()
    {
        string? value = "42";
        const int count = 84;
        if (value?.Length == count)
        {
            Console.WriteLine(value.Length); // no warning
        }
    }
}
```

cc: @dotnet/nullablefc
